controller/registry/tests.py

```diff
@@ -0,0 +1,89 @@
+"""
+Unit tests for the Deis registry app.
+
+Run the tests with "./manage.py test registry"
+"""
+
+import unittest
+try:
+    from unittest import mock
+except ImportError:
+    import mock
+
+from django.conf import settings
+from rest_framework.exceptions import PermissionDenied
+from registry.dockerclient import DockerClient
+
+
+@mock.patch('docker.Client')
+class DockerClientTest(unittest.TestCase):
+    """Test that the client makes appropriate Docker engine API calls."""
+
+    def setUp(self):
+        settings.REGISTRY_HOST, settings.REGISTRY_PORT = 'localhost', 5000
+
+    def test_publish_release(self, mock_client):
+        self.client = DockerClient()
+        self.client.publish_release('ozzy/embryo:git-f2a8020',
+                                    {'POWERED_BY': 'Deis'}, 'ozzy/embryo:v4', True)
+        self.assertTrue(self.client.client.pull.called)
+        self.assertTrue(self.client.client.tag.called)
+        self.assertTrue(self.client.client.build.called)
+        self.assertTrue(self.client.client.push.called)
+        # Test that blacklisted image names can't be published
+        with self.assertRaises(PermissionDenied):
+            self.client.publish_release(
+                'deis/controller:v1.11.1', {}, 'deis/controller:v1.11.1', True)
+        with self.assertRaises(PermissionDenied):
+            self.client.publish_release(
+                'localhost:5000/deis/controller:v1.11.1', {}, 'deis/controller:v1.11.1', True)
+
+    def test_build(self, mock_client):
+        # test that self.client.build was called with proper arguments
+        self.client = DockerClient()
+        self.client.build('ozzy/embryo:git-f3a8020', {'POWERED_BY': 'Deis'}, 'ozzy/embryo', 'v4')
+        docker_build = self.client.client.build
+        self.assertTrue(docker_build.called)
+        args = {"rm": True, "tag": u'localhost:5000/ozzy/embryo:v4', "stream": True}
+        kwargs = docker_build.call_args[1]
+        self.assertDictContainsSubset(args, kwargs)
+        # test that the fileobj arg to "docker build" contains a correct Dockerfile
+        f = kwargs['fileobj']
+        self.assertEqual(f.read(), "FROM ozzy/embryo:git-f3a8020\nENV POWERED_BY='Deis'")
+        # Test that blacklisted image names can't be built
+        with self.assertRaises(PermissionDenied):
+            self.client.build('deis/controller:v1.11.1', {}, 'deis/controller', 'v1.11.1')
```

Comment:
> We should just blacklist apps being created with these names.

I can move the `_check_blacklist` method up to a module-level function and call it at `deis create` time.

> We should also blacklist anyone from creating a deis user so we fail early

That could be implemented in serializers.py easily. But I agree banning a user named "deis" would, strictly speaking, not be backward-compatible, nor would the previous change. We may want to do so anyway because the Docker deprecation event forces us to implement this PR (or something like it).